File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        n = len(nums)
        window = set()
        for i,num in enumerate(nums):
            # check if r in the set
            if num in window:
                return True
            window.add(num)
            if len(window)>k:
                window.remove(nums[i-k])
        return False
        

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.cpp
class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        unordered_set<int> window;
        int n = nums.size();
        int l=0;

        for(int r=0;r<n;r++){
            if(r-l>k){
                window.erase(nums[l]);
                l++;
            }
            if(window.find(nums[r]) != window.end()){
                return true;
            }
            window.insert(nums[r]);
        }
        return false;
    }
};